### PR TITLE
Inspect correct function in wrap_init

### LINKED ISSUE #12564: Inheritance of Data Loaders
## 🐛 Bug

This is the official bug report of my comment earlier (see https://github.com/PyTorchLightning/pytorch-lightning/pull/10680#issuecomment-1084488190).

```python
import pytorch_lightning as pl
import torch


class Model(pl.LightningModule):
    def __init__(self, in_channels):
        super().__init__()
        self.lin = torch.nn.Linear(in_channels, 1)

    def forward(self, x):
        return self.lin(x)

    def training_step(self, x, batch_idx):
        loss = self(x).mean()
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=0.01)


class MyBaseDataLoader(torch.utils.data.DataLoader):
    pass


class MyDataLoader(MyBaseDataLoader):
    def __init__(self, data: torch.Tensor, *args, **kwargs):
        self.data = data
        super().__init__(range(data.size(0)), *args, **kwargs,
                         collate_fn=self.sample_fn)

    def sample_fn(self, indices):
        assert isinstance(self.data, torch.Tensor)
        index = torch.tensor(indices)
        return self.data[index]


class DataModule(pl.LightningDataModule):
    def __init__(self, data: torch.Tensor):
        super().__init__()
        self.data = data

    def train_dataloader(self):
        return MyDataLoader(self.data, batch_size=32)


model = Model(in_channels=32)
datamodule = DataModule(torch.randn(100, 32))

trainer = pl.Trainer(accelerator='gpu', devices=1, max_epochs=10)
trainer.fit(model, datamodule)
```

which crashes with
```
  File "/home/matthias/miniconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 530, in __next__
    data = self._next_data()
  File "/home/matthias/miniconda3/lib/python3.9/site-packages/torch/utils/data/dataloader.py", line 570, in _next_data
    data = self._dataset_fetcher.fetch(index)  # may raise StopIteration
  File "/home/matthias/miniconda3/lib/python3.9/site-packages/torch/utils/data/_utils/fetch.py", line 52, in fetch
    return self.collate_fn(data)
  File "/home/matthias/github/pytorch_geometric/alpha/PL.py", line 32, in sample_fn
    assert isinstance(self.data, torch.Tensor)
AssertionError

```

Notably, the code runs through for `pytorch-lightning==1.5.10` and when I let `MyDataLoader` inherit directly from `torch.utils.data.DataLoader`.

I personally think this issue is related to overriding both `__init__` from `MyBaseDataLoader` and `MyDataLoader` [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/data.py#L358).

- PyTorch Lightning Version (e.g., 1.5.0): 1.6.0
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9): 3.9
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3

cc @justusschock @awaelchli @ninginthecloud

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

In the function `_wrap_init`, the function of the live object was inspected, which was incorrect in the case of multiple inheritance, as the signature was always of the one function only. This PR inspects the function that should be actually invoked.  

Fixes #12564

Extra test case, that copies the behavior from #12564 is created.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @borda @akihironitta @carmocca

### CODE REVIEW

File: pytorch_lightning/utilities/data.py [hunk 1]
```diff
@@ -321,11 +321,20 @@ def _wrap_init(init: Callable) -> Callable:
 
     @functools.wraps(init)
     def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
-        params = dict(inspect.signature(obj.__init__).parameters)
-        params.pop("args", None)
-        params.pop("kwargs", None)
+        # We need to inspect init function, as inspecting live function `obj.__init__``
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:324]:
```suggestion
        # We need to inspect `init`, as inspecting `obj.__init__`
```

File: pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -321,11 +321,20 @@ def _wrap_init(init: Callable) -> Callable:
...
+        # can lead to inspecting the wrong function with complicated inheritance
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:325]:
```suggestion
        # can lead to inspecting the wrong function with multiple inheritance
```

File: tests/utilities/test_data.py [hunk 3]
```diff
@@ -161,20 +161,35 @@ def __init__(self, attribute1, *args, **kwargs):
...
 
     with _replace_dataloader_init_method():
-        dataloader = DataLoaderSubclass2("attribute1", "attribute2", dataset=range(4), batch_size=2)
-        assert dataloader.attribute1 == "attribute1"
+        dataloader = DataLoaderSubclass2("attribute2", dataset=range(4), batch_size=2)
+        assert dataloader.attribute1 == "attribute2-2"
         assert dataloader.attribute2 == "attribute2"
 
+    # Failing test case from issue 12564
+    class MyBaseDataLoader(DataLoader):
```
Comment (carmocca) [inline tests/utilities/test_data.py:179]: Awesome test addition!

**Review by carmocca — APPROVED** [review]
Nice!

File: pytorch_lightning/utilities/data.py [hunk 4]
```diff
@@ -321,11 +321,18 @@ def _wrap_init(init: Callable) -> Callable:
...
-        params.pop("args", None)
-        params.pop("kwargs", None)
+        # We need to inspect `init`, as inspecting `obj.__init__`
+        # can lead to inspecting the wrong function with multiple inheritance
+        params = inspect.signature(init).parameters
+
+        param_names = [
+            param.name
+            for param in params.values()
+            if param.name != "self" and param.kind != param.VAR_POSITIONAL and param.kind != param.VAR_KEYWORD
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/data.py:331]:
```suggestion
            if param.name != "self" and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
```
nit

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
nice.

**Comment by otaj** [comment]
@carmocca Done in #12813
